src/promise.rs

```diff
@@ -0,0 +1,338 @@
+                    new_promise_for_catch.reject(reason);
+                }));
+            }
+            PromiseState::Fulfilled(value) => {
+                let result = f(value.clone());
+                new_promise.fulfill(result);
+            }
+            PromiseState::Rejected(reason) => new_promise.reject(reason.clone()),
+        }
+        new_promise
+    }
+
+    pub fn catch<F>(self: &Rc<Self>, f: F) -> Rc<Self>
+    where
+        F: FnOnce(String) + 'static,
+    {
+        match &*self.state.borrow() {
+            PromiseState::Pending => *self.catch_callback.borrow_mut() = Some(Box::new(f)),
+            PromiseState::Fulfilled(_) => {}
+            PromiseState::Rejected(reason) => f(reason.clone()),
+        }
+        self.clone()
+    }
+
+    pub fn all_of(promises: Vec<Rc<Self>>) -> Rc<Promise<Vec<T>>> {
+        let next_promise = Promise::new();
+
+        if promises.is_empty() {
+            next_promise.fulfill(vec![]);
+            return next_promise;
+        }
+
+        let total = promises.len();
+        let results = Rc::new(RefCell::new(vec![None; total]));
+        let remaining = Rc::new(RefCell::new(total));
+        let rejected = Rc::new(RefCell::new(false));
+
+        for (i, promise) in promises.iter().enumerate() {
+            let next_promise_clone = next_promise.clone();
+            let next_promise_clone_for_catch = next_promise.clone();
+            let results_clone = results.clone();
+            let remaining_clone = remaining.clone();
+            let rejected_clone = rejected.clone();
+            let rejected_clone_for_catch = rejected.clone();
+            promise
+                .then(move |result| {
+                    if *rejected_clone.borrow() {
+                        return;
+                    }
+                    results_clone.borrow_mut()[i] = Some(result);
+                    *remaining_clone.borrow_mut() -= 1;
+
+                    if *remaining_clone.borrow() == 0 {
+                        let final_results: Vec<T> = results_clone
+                            .borrow_mut()
+                            .iter_mut()
+                            .map(|res| res.take().unwrap())
+                            .collect();
+                        next_promise_clone.fulfill(final_results);
+                    }
+                })
+                .catch(move |reason| {
+                    if !*rejected_clone_for_catch.borrow() {
+                        *rejected_clone_for_catch.borrow_mut() = true;
+                        next_promise_clone_for_catch.reject(reason.clone());
+                    }
+                });
+        }
+        next_promise
+    }
+
+    pub fn any_of(promises: Vec<Rc<Self>>) -> Rc<Promise<T>> {
+        let next_promise = Promise::new();
+        let total = promises.len();
+        let remaining = Rc::new(RefCell::new(total));
+        let first_error = Rc::new(RefCell::new(None)); // 用来保存第一个错误
```

Comment:
Please stick to English, thanks!

Reply:
Sorry, my mistake, I will remove all Chinese comments